Initial nullable api support (for Kotlin)

checketts opened the pull request:
This adds initial Nullability support.

This PR is a work in progress as I add `Nullable` to any places where we return any nulls.

I'm mimicking the Spring annotation/approach. Any recommendations or tips @sdeleuze?

jkschneider commented:
Was there a talk on this that I could listen to? Seems like you are safe to use optional dependencies via meta annotations that just add a layer of indirection?

jkschneider commented:
Addresses #123

jkschneider commented on micrometer-core/src/main/java/io/micrometer/core/lang/Nullable.java:
```diff
@@ -0,0 +1,49 @@
+/**
+ * A common annotation to declare that annotated elements can be {@code null} under
+ * some circumstance. Leverages JSR 305 meta-annotations to indicate nullability in Java
+ * to common tools with JSR 305 support and used by Kotlin to infer nullability of Spring API.
```
Reference to Spring.

jkschneider commented on micrometer-core/src/main/java/io/micrometer/core/lang/NonNullFields.java:
```diff
@@ -0,0 +1,47 @@
+ * non-nullable by default for a given package.
+ *
+ * <p>Leverages JSR-305 meta-annotations to indicate nullability in Java to common
+ * tools with JSR-305 support and used by Kotlin to infer nullability of Spring API.
```
Reference to Spring

jkschneider commented on micrometer-core/build.gradle:
```diff
@@ -8,6 +8,10 @@ apply plugin: 'me.champeau.gradle.jmh'
 
 dependencies {
+    // JSR-305 only used for non-required meta-annotations
+    compileOnly("com.google.code.findbugs:jsr305:3.0.2")
```
Use optional instead of provided?

checketts commented:
Unfortunately the talks where it was mentioned weren't exhaustive. They typically had a slide that said 'just add package_info.java' and went on. So that is why I inspected how Spring did it and mimicked them. That is also why I used `compileOnly` instead of making it `optional`, since that is how Spring approached it.

The *Idiomatic Interop* talk got me the most excited: https://youtu.be/zlggmRMadsU?t=25m49s (I tried to link to the slide part (around 25:40).

I believe it was also covered in *The Road to KotlinTown* https://youtu.be/OmwjrVawHqA

checketts commented:
I removed the Spring references and switched to the optional compile dep. Kotlin seems to still respect the non-null setting that way.

As I'm checking each class to see if they return null, how should I approach nullability in codebase?

For example with the `AbstractMeter` class, is a null is passed in as the ID then it could return a null from `getId()`, should I begin adding in NonNull assertions?

```
public abstract class AbstractMeter implements Meter {
    private final Meter.Id id;

    public AbstractMeter(Id id) {
        this.id = id;
    }

    @Override
    public Id getId() {
        return id;
    }
}
```

jkschneider commented:
> For example with the AbstractMeter class, if a null is passed in as the ID then it could return a null from getId(), should I begin adding in NonNull assertions in the constructor?

Yes.

checketts commented on micrometer-core/src/main/java/io/micrometer/core/instrument/DistributionSummary.java:
```diff
@@ -133,7 +137,7 @@ public Builder publishPercentileHistogram() {
          * Atlas' {@code :percentiles}).
          */
-        public Builder publishPercentileHistogram(Boolean enabled) {
+        public Builder publishPercentileHistogram(boolean enabled) {
```
Switched to primitives to avoid null checks

checketts commented on micrometer-core/src/main/java/io/micrometer/core/instrument/HistogramSnapshot.java:
```diff
@@ -98,11 +101,11 @@ public String toString() {
         buf.append(", max=");
         buf.append(max);
-        if (percentileValues.length > 0) {
+        if (percentileValues != null && percentileValues.length > 0) {
```
Uncovered some potential null pointer exceptions

checketts commented on micrometer-core/src/main/java/io/micrometer/core/instrument/MeterRegistry.java:
```diff
@@ -111,16 +114,17 @@ protected String getConventionName(Meter.Id id) {
     protected abstract TimeUnit getBaseTimeUnit();
 
     private String getBaseTimeUnitStr() {
-        if (getBaseTimeUnit() == null) {
```
All registries don't return null. Removed the check to clean up the API

checketts commented on micrometer-core/src/main/java/io/micrometer/core/instrument/MeterRegistry.java:
```diff
@@ -597,7 +610,9 @@ public More more() {
 
-    private Meter getOrCreateMeter(HistogramConfig config, BiFunction<Id, HistogramConfig, Meter> builder, Id mappedId) {
+    private Meter getOrCreateMeter(@Nullable HistogramConfig config,
+                                   BiFunction<Id, /*Nullable Generic*/ HistogramConfig, Meter> builder,
```
I'm not certain how to annotate the generic to indicate that `HistogramConfig` could be null. Perhaps we shouldn't allow it to be `null` ever?

checketts commented on micrometer-core/src/main/java/io/micrometer/core/instrument/histogram/HistogramConfig.java:
```diff
@@ -57,50 +58,52 @@ public HistogramConfig merge(HistogramConfig parent) {
 
     public boolean isPublishingHistogram() {
-        return percentileHistogram || sla.length > 0;
+        return (percentileHistogram != null && percentileHistogram) || (sla != null && sla.length > 0);
```
Fixed potential NPE

checketts commented on micrometer-core/src/main/java/io/micrometer/core/instrument/histogram/HistogramConfig.java:
```diff
@@ -57,50 +58,52 @@ public HistogramConfig merge(HistogramConfig parent) {
 
-        for (long slaBoundary : sla) {
-            buckets.add(slaBoundary);
+        if(sla != null) {
```
fixed potential NPE

checketts commented on micrometer-core/src/main/java/io/micrometer/core/instrument/util/Assert.java:
```diff
@@ -0,0 +1,19 @@
+     * @param field the exception message to use if the assertion fails
+     * @throws IllegalArgumentException if the object is {@code null}
+     */
+    public static void notNull(@Nullable Object object, String field) {
```
This is the assertion check code

checketts commented on micrometer-core/src/main/java/io/micrometer/core/instrument/Gauge.java:
```diff
@@ -46,10 +49,13 @@ default Type type() {
 
         private Builder(String name, T obj, ToDoubleFunction<T> f) {
+            Assert.notNull(name,"name");
+            Assert.notNull(obj,"gauge object");
```
This change is causing the dropwizard null gauge test to fail. Should `null`s be allowed for `Gauge`s @jkschneider ?

jkschneider replied:
It is incorrect. The obj can become dereferenced and, since our gauge implementations maintain a weak reference to it, thus `null`. I think we want to maintain that behavior.

In general, I'm not keen on adding assert code anywhere in the main source set.

checketts replied:
> I'm not keen on adding assert code anywhere in the main source set.

Sorry I thought you had requested that I add the asserts. I'm fine removing them. Or by `main source set` are you saying to leave them out of the common code paths but leave them in other places like binders?

jkschneider replied:
> should I begin adding in NonNull assertions in the constructor?

Sorry, I thought you meant adding `@NonNull` annotations on constructor parameters.
